File: leetcode/problems/language-specific/go/167.two-sum-ii-input-array-is-sorted.go
/*
 * @lc app=leetcode id=167 lang=golang
 *
 * [167] Two Sum II - Input Array Is Sorted
 */

// @lc code=start

// * Two Pointer Solution | O(n) Time | O(1) Space

func twoSum(numbers []int, target int) []int {
	numbersSize := len(numbers)
	res := []int{0, 0}
	if numbersSize <= 1 {
		return res
	}

	left, right := 0, numbersSize-1
	for left < right {
		totalSum := numbers[left] + numbers[right]
		if totalSum == target {
			res[0] = left + 1
			res[1] = right + 1
			return res
		} else if totalSum < target {
			left++
		} else {
			right--
		}
	}

	return res
}

// @lc code=end

File: leetcode/problems/language-specific/python/167.two-sum-ii-input-array-is-sorted.py
#
# @lc app=leetcode id=167 lang=python3
#
# [167] Two Sum II - Input Array Is Sorted
#

# @lc code=start

# * Two Pointers and Binary Search Solution | O(n) Time | O(1) Space
# * n -> The length of numbers array


class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        numbers_len = len(numbers)
        left, right = 0, numbers_len - 1
        while left < right:
            cur_sum = numbers[left] + numbers[right]
            if cur_sum == target:
                return [left + 1, right + 1]
            elif cur_sum < target:
                left += 1
            else:
                right -= 1


# @lc code=end
